### ISSUE TITLE
".." entry in folders

### ISSUE THREAD

**Issue by mherrmann** [fman-community/fman#35]
Pros:
------
 + Intuitive
 + Can be used with mouse
 + It's sometimes useful to be able to copy the path of the current directory to the clipboard. F11 on ".." would enable this. However, without "..", you can still go up (Backspace) and then press F11.
 + In Total Commander, when you Shift+F6 "..", you can directly enter the path of a directory to navigate to. I find this quite intuitive.

Cons:
------
 - Clutters up UI; Makes fman less "beautiful".
 - When you enter a folder right now, the first file/directory in that folder is automatically selected. If that happens to be the file/directory you want to open, you save a key stroke! If ".." were there, at the top of the list and thus automatically selected, you would always have to go down/type something at least once to interact with the directory contents
 - It's a little more difficult to implement
 - I would argue not having the ".." there has the benefit of educating the user. One user learned about the Backspace functionality because ".." isn't there. From now on, he knows how to use Backspace - and will be quicker in the process because it is just a single key to type.

Since there are so many pros and cons, it will be best to simply make this configurable so each user can set it up the way (s)he likes.

**Comment by vovcacik** [comment]
The two dots `..` refer to the parent directory. So F11 on it should not copy the path of the current directory. You would need `.` placeholder for it.

**Comment by mherrmann** [comment]
Good point.

**Comment by ar-jan** [comment]
I just tried fman for the first time, and I missed this. I'd suggest `Alt+Up` should also navigate to parent dir.

**Comment by mherrmann** [comment]
Thank you for the feedback / suggestion @ar-jan. I've created a [separate issue](https://github.com/fman-users/fman/issues/329) for `Alt-Up`.

**Comment by zorzysty** [comment]
In my humble opinion Alt + Up is sufficient solution. 
If you ever decide to add '..', please make it optional. For me it would negatively affect usability.

**Comment by mherrmann** [comment]
No worries, if I implement it, it will be optional.

**Comment by zorzysty** [comment]
Good to hear :)
What I would welcome with open arms though would be making active path on top clickable so upon clicking it would take you to the clicked folder (like in windows explorer).

**Comment by mherrmann** [comment]
You mean when (say) you're at `C:\Users\Michael\dev` and click on `Michael`, then fman takes you to `C:\Users\Michael`?

**Comment by zorzysty** [comment]
Yes, exactly that.

**Comment by mherrmann** [comment]
Okay. Can I ask you to open a separate issue for this? It's not really part of this one.

**Comment by zorzysty** [comment]
Done: [#472](https://github.com/fman-users/fman/issues/472)

**Comment by mherrmann** [comment]
Thank you!

**Comment by pravic** [comment]
Also experience from TC: when you enter an empty directory, you can go back simply by pressing _Enter_ (the same way as you've entered). 

Here, I have a "No file is selected!" alert, which is confusing at first.

**Comment by mherrmann** [comment]
I highly assume you do know the <kbd>Backspace</kbd> shortcut for going up a directory @pravic. Just wanted to mention it to make sure.

**Comment by pravic** [comment]
Of course. I was talking about "user experience" (i.e. observable behaviour which can be expectable in similar programs).

**Comment by pravic** [comment]
Yes, I confirm it: having to press one button (Enter) instead of two (Enter+Backspace) in order to go back from an empty directory is priceless.

@mherrmann Is there a way to achieve this with a plugin? Or it can be done only by fman itself?

**Comment by mherrmann** [comment]
Yes, you could implement this as a plugin. But I'm curious: Why do you have to press <kbd>Enter</kbd>+<kbd>Backspace</kbd> to go back? Or do you mean the <kbd>Enter</kbd> is to open the empty directory in the first place?

If you want me to give you pointers on how to write the plugin, just let me know. Should be pretty easy.

**Comment by pravic** [comment]
>  Or do you mean the Enter is to open the empty directory in the first place?

This. How else would you enter a directory? :)

As for instructions - yes, they would be helpful.

**Comment by mherrmann** [comment]
Yes you are right, it was not the smartest question ;-)

Okay! [Create a plugin](https://fman.io/docs/plugins-introduction) with the following Python code:

    from fman import DirectoryPaneListener

    class GoUpOnEnterInEmptyDirectory(DirectoryPaneListener):
        def on_command(self, command, args):
            if command == 'open':
                if not self.pane.get_file_under_cursor():
                    return 'go_up', {}

(You don't have to create a `Key Bindings.json` file.)

If that leave something unanswered, let me know.

**Comment by pravic** [comment]
Works like a charm. Thanks!

**Comment by mherrmann** [comment]
Happy to hear it!

**Comment by dcsoft4** [comment]
My opinion is that Backspace is not so intuitive to go to the parent directory.  A browser uses Alt+Left arrow to go Back, and I think people are more used to that.  Something visual is needed.

**Comment by mherrmann** [comment]
The question is: Do I want to optimize fman for first time use, or for the power user who already knows it. The double dots ".." are visual and may thus be more intuitive for first time users. But power users, who should use the keyboard most of the time because that's what fman is optimized for, will no longer wonder "how do I go up a directory?". For them, the double dots require screen estate for (imho) little added benefit.

In my eyes, it's similar to the <kbd>F3</kbd>, <kbd>F4</kbd>, ... button row which other file managers display at the bottom:

![image](https://user-images.githubusercontent.com/1076393/49276069-96080f00-f47d-11e8-8a45-c56d127b011c.png)

Yes it's nice for first-time users. But totally unnecessary for heavy users.

I strive to help first time users eg. with the [tutorial](https://fman.io/docs/tutorial).

**Comment by proudot** [comment]
Hi there, this is one of the few features preventing me from fully switching to fman. Other software can use mouse heavily and having to switch from mouse-oriented work to keyboard-only work is not productive. The ability to navigate with the mouse would be game changing for my usecase.

**Comment by mherrmann** [comment]
Thank you for the feedback @proudot. One of fman's strengths (imho) is that it is laser-focused on doing one thing, and that thing well. This thing is efficient file management with the keyboard. fman really isn't meant to be used with the mouse, and never will be. Sure, you can click around in fman because _sometimes_ you may need it. But adding the `..` entry - which is then _always_ visible and takes up screen estate - is too high a cost for that occasional use. Sorry.

**Comment by proudot** [comment]
Ok, and how about a customizable shortcut mapped to a mouse click, such as the third button ?

**Comment by eandmsz** [comment]
I agree with @proudot . This would be a great addition to fman, so users who have decades of muscle memory can switch easier. Make it configurable and disable it by default if you want to show a "clean" interface for the new users.
It would be also good to have the mouse side "back" button to move up to parent.
More options for an often used functionality doesn't make the software more cluttered.

**Comment by vfxdev** [comment]
This issue now has 30 votes and @eandmsz brings up a great way to make everyone happy. This is the single thing holding me back from making fman my default file manager on Linux. People downloading a NC/TC clone expect the basic functionality to be identical, otherwise they'll turn somewhere else.

Thanks!

**Comment by strayge** [comment]
I've added ".." entries by plugin - [FmanDotEntries](https://github.com/strayge/FmanDotEntries).  
Work for me at Win & Mac.

BTW I didn't found any API to add new entries so it a bit clutters.  
If anyone knows best way - PRs welcomed.

**Comment by mherrmann** [comment]
Epic. Thank you @strayge!
